### PR TITLE
[ISSUE #10522] Reduce Remoting header encoding allocation via FastCodesHeader/RocketMQSerializable optimizations

### LINKED ISSUE #10522: [Enhancement] Reduce Remoting header encoding allocation via FastCodesHeader/RocketMQSerializable optimizations
### Before Creating the Enhancement Request

- [x] I have confirmed that this should be classified as an enhancement rather than a bug/feature.

### Summary

Enhance the `FastCodesHeader` interface and `RocketMQSerializable` utility to reduce allocation and CPU overhead during Remoting header encoding/decoding.

### Motivation

In high-throughput scenarios, every RPC request/response goes through the Remoting header encode/decode path. Current implementation has several inefficiencies:

1. `FastCodesHeader.writeIfNotNull` always converts numeric values (`Long`/`Integer`) to `String` via `value.toString()`, creating transient String objects.
2. `RocketMQSerializable.readStr` allocates a new `String` even for single-byte ASCII values.
3. `RocketMQSerializable.mapDeserialize` pre-allocates a `HashMap` with capacity 128, which is oversized for typical headers and wastes memory.
4. There is no direct way to write primitive `long`/`int` values to the ByteBuf without first boxing them into `Long`/`Integer`.

These small overheads are amplified by millions of RPCs per second in production.

### Describe the Solution You'd Like

1. Add `writeDecimalLong(ByteBuf, long)` and `writeDecimalInt(ByteBuf, int)` to `RocketMQSerializable` to write primitive decimal numbers directly to ByteBuf without String allocation.
2. Add `writeLong(ByteBuf, String, long)` and `writeInt(ByteBuf, String, int)` helpers to `FastCodesHeader`.
3. Optimize `FastCodesHeader.writeIfNotNull` to route `Long`/`Integer` values through the new primitive writers.
4. Add a `SINGLE_BYTE_STRINGS` cache for ASCII single-byte strings in `RocketMQSerializable.readStr`.
5. Reduce the initial capacity of `HashMap` in `mapDeserialize` from 128 to 24.

All changes are additive and keep existing method signatures backward compatible.

### Describe Alternatives You've Considered

- Use `String.valueOf` + `writeStr`: still allocates String objects.
- Use `ByteBuf.writeAscii` for numbers: Netty's API does not provide direct decimal writing with length prefix required by RocketMQ protocol.
- Keep `HashMap(128)` in `mapDeserialize`: wastes memory for small maps; 24 is sufficient for typical headers.

### Additional Context

This is part of a larger effort to reduce Broker allocation rate. Related previous PRs: #10443, #10491, #10485, #10514, #10513, #10517.

### PR DESCRIPTION
**Pull request by wang-jiahua**
### Which Issue(s) This PR Fixes

Fixes #10522

### Brief Description

Enhance `FastCodesHeader` interface and `RocketMQSerializable` utility to reduce allocation and CPU overhead in the Remoting header encoding/decoding hot path.

**`FastCodesHeader.java`**
- Optimize `writeIfNotNull` to route `Long`/`Integer` values through `writeDecimalLong`/`writeDecimalInt` instead of `value.toString()` + `writeStr`, eliminating transient `String` allocation.
- Add `writeLong(ByteBuf, String, long)` and `writeInt(ByteBuf, String, int)` helper methods for direct primitive encoding.

**`RocketMQSerializable.java`**
- Add `writeDecimalLong(ByteBuf, long)` and `writeDecimalInt(ByteBuf, int)` — write decimal numbers directly to `ByteBuf` without `String.valueOf()` allocation.
- Add `SINGLE_BYTE_STRINGS` cache (128 entries) — `readStr` returns cached `String` for single-byte ASCII values instead of creating new `String` via `readCharSequence`.
- Right-size `mapDeserialize` initial `HashMap` capacity from 128 to 24 — typical headers have 5-15 entries, 128 causes unnecessary memory waste.

All changes are additive — no existing method signatures are modified.

### How Did You Test This Change?

**1. Unit tests**

```
mvn test -pl remoting -Dtest=RocketMQSerializableTest,FastCodesHeaderTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0
```

**2. A/B benchmark** (256 threads, 1KB sync send, Dragonwell JDK 21, 6g heap)

| Metric | Baseline (develop) | Optimized (R1) | Change |
|--------|-------------------|----------------|--------|
| TPS | ~196k | ~236k | **+20.6%** |
| p99 RT | 3ms | 2ms | **-33%** |
| Average RT | 1.307ms | 1.085ms | **-17.0%** |
| Max RT | 594ms | 351ms | **-40.9%** |
| Young GCs (total) | 336 | 304 | **-9.5%** |

**3. Commercial compatibility**

- Pure additive changes — no existing API signatures modified.
- No commercial classes implement `FastCodesHeader`.
- Commercial `YitianRemotingCommand` calls `RocketMQSerializable.rocketMQProtocolDecode/Encode` which are unchanged.

### CODE REVIEW

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 1]
```diff
@@ -28,6 +28,74 @@
...
+            numDigits++;
+            temp /= 10;
+        }
+        buf.writerIndex(digitStart + numDigits);
+        temp = value;
+        for (int i = numDigits - 1; i >= 0; i--) {
+            buf.setByte(digitStart + i, (byte) ('0' + (int) (temp % 10)));
+            temp /= 10;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:99]: Good catch. Added `buf.ensureWritable(numDigits)` before advancing the writer index in `writePositiveDigits`. Fixed in latest push.

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 2]
```diff
@@ -28,6 +28,74 @@
...
+                    buf.writeByte('8');
+                    buf.setInt(lenIndex, buf.writerIndex() - start);
+                    return;
+                }
+                value = -value;
+            }
+            writePositiveDigits(buf, value);
+        }
+        buf.setInt(lenIndex, buf.writerIndex() - start);
+    }
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:82]: The duplication is intentional. Each method handles a different primitive type with type-specific MIN_VALUE edge cases (`Long.MIN_VALUE` vs `Integer.MIN_VALUE`). Consolidating would require either autoboxing or a shared `long` implementation with an `int` wrapper that still needs its own MIN_VALUE check. The current approach keeps each method self-contained and avoids unnecessary type conversions.

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 3]
```diff
@@ -52,6 +120,13 @@ private static String readStr(ByteBuf buf, boolean useShortLength, int limit) th
...
         }
+        if (len == 1) {
+            byte b = buf.readByte();
+            if (b >= 0) {
+                return SINGLE_BYTE_STRINGS[b];
+            }
+            return new String(new byte[]{b}, StandardCharsets.UTF_8);
+        }
         CharSequence cs = buf.readCharSequence(len, StandardCharsets.UTF_8);
         return cs == null ? null : cs.toString();
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:132]: Good catch. Changed to use the existing `CHARSET_UTF8` constant for consistency. Fixed in latest push.

File: remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java [hunk 4]
```diff
@@ -231,7 +306,7 @@ public static RemotingCommand rocketMQProtocolDecode(final ByteBuf headerBuffer,
 
     public static HashMap<String, String> mapDeserialize(ByteBuf byteBuffer, int len) throws RemotingCommandException {
 
-        HashMap<String, String> map = new HashMap<>(128);
+        HashMap<String, String> map = new HashMap<>(24);
```
[the remark this answers is not in the document]
Comment (wang-jiahua) [reply remoting/src/main/java/org/apache/rocketmq/remoting/protocol/RocketMQSerializable.java:311]: Added a comment explaining the rationale: typical RocketMQ headers have 5-15 entries, and 24 is the next power-of-2 above 15 × 0.75 (default load factor), which avoids resize. Fixed in latest push.

**Review by RockteMQ-AI — COMMENTED** [review]
## Review by github-manager-bot

### Summary
This PR optimizes the Remoting header encoding/decoding hot path by eliminating transient `String` allocations for numeric values and adding a single-byte ASCII string cache. The changes are purely additive with impressive benchmark results (+20.6% TPS, -33% p99 RT).

### Findings

- **[Positive]** `FastCodesHeader.java:41-47` — Routing `Long`/`Integer` through `writeDecimalLong`/`writeDecimalInt` instead of `value.toString()` + `writeStr` is a clean and effective allocation elimination. The `instanceof` branching cost is negligible compared to the allocation savings on the hot path.

- **[Positive]** `RocketMQSerializable.java` `writeDecimalLong`/`writeDecimalInt` — The `Long.MIN_VALUE` / `Integer.MIN_VALUE` edge cases are correctly handled by dividing by 10 first and appending the last digit, avoiding the overflow that `-value` would cause. Well done.

- **[Positive]** `SINGLE_BYTE_STRINGS` cache — 128-entry ASCII cache is a good trade-off: ~5-6KB static memory vs. avoiding `new String()` for the most common single-byte header values. The `readStr` fast path (`len == 1 && b >= 0`) is correct for UTF-8 single-byte range.

- **[Positive]** `mapDeserialize` capacity 128 → 24 — Good observation. Typical headers have 5-15 entries; 128 was wasteful. With load factor 0.75, capacity 24 supports 18 entries before resize, which covers the common case.

- **[Info]** `FastCodesHeader.java:41` — Consider reordering the `instanceof` checks to test `String` first (most common header value type), then `Long`, then `Integer`. This avoids two unnecessary type checks for the majority of calls. Minor optimization, non-blocking.

```java
if (value instanceof String) {
    RocketMQSerializable.writeStr(out, false, (String) value);
} else if (value instanceof Long) {
    RocketMQSerializable.writeDecimalLong(out, (Long) value);
} else if (value instanceof Integer) {
    RocketMQSerializable.writeDecimalInt(out, (Integer) value);
} else {
    RocketMQSerializable.writeStr(out, false, value.toString());
}
```

- **[Info]** `writePositiveDigits` uses a two-pass approach (count digits, then write). An alternative is a pre-allocated digit lookup table indexed by value, but the current approach is clean and the performance difference is negligible for numbers up to 19 digits.

### Compatibility
- All changes are additive — no existing method signatures modified.
- New `default` methods in `FastCodesHeader` interface do not break existing implementations.
- `SINGLE_BYTE_STRINGS` is a private static field with no external visibility.
- Wire format is unchanged: `writeDecimalLong`/`writeDecimalInt` produce the same byte output as `writeStr(out, false, String.valueOf(value))`.

### Tests
- Unit tests cover `RocketMQSerializable` and `FastCodesHeader` directly.
- Benchmark data is well-documented with clear before/after comparison.
- Consider adding an edge-case test for `writeDecimalLong(Long.MIN_VALUE)` and `writeDecimalInt(Integer.MIN_VALUE)` to prevent regression of the overflow handling.

---
*Automated review by github-manager-bot*

**Review by RongtongJin — APPROVED** [review]
